### ISSUE TITLE
Highlight positions do not update to the latest location of UIViews

### ISSUE THREAD

**Issue by shirakaba** [Tawa/TNTutorialManager#13]
# Summary

In the following example, I adjust the horizontal layout margins of a UIStackView during `viewWillTransition()` between portrait/landscape. However, TNTutorialManager only seems to take the initial dimensions of the UIView for a given orientation, and – although it stretches those initial highlight positions correctly between rotations – it does not update the highlight positions based on their latest frame, which the developer may need to adjust upon rotation.

It would be nice if TNTutorialManager could update the highlight position to the latest frame at least on each rotation, and ideally at regular intervals (to allow tracking an animated target, like a fish moving in a game).

# Real example

I am trying to highlight UIBarButtons on the toolbar. However, UIBarButtonItems do not extend UIView, so I can't directly pass then to `tutorialViewsToHighlight()`. To circumvent this, I am thus using a UIStackView filled with UIViews to line up exactly with each of the UIBarButtonItems on the toolbar. I found that the stackview must change its margins upon each rotation to line its UIViews up with the UIBarButtonItems properly.

## Code

```Swift
import UIKit
import Foundation
import TNTutorialManager

class ViewController: UIViewController, WKScriptMessageHandler {
    var tm: TNTutorialManager!
	let toolbarMimicStack: UIStackView = UIStackView()

	override func loadView() {
		super.loadView()

		// Reset tutorial to starting index (tutorialIdentifier is "mainVC")
		UserDefaults.standard.removeObject(forKey: "TNTutorial" + "mainVC")
        tm = TNTutorialManager(delegate: self, blurFactor: 0.1)

        let spacer: UIBarButtonItem = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
        let dummyItem: UIBarButtonItem = UIBarButtonItem(title: "三", style: UIBarButtonItemStyle.plain, target: self, action: nil)
		toolbarItems = [dummyItem, spacer, dummyItem, spacer, dummyItem, spacer, dummyItem]

		// Initialise stackview layout
		toolbarMimicStack.isLayoutMarginsRelativeArrangement = true
		toolbarMimicStack.axis = .horizontal
		toolbarMimicStack.alignment = .fill
		toolbarMimicStack.distribution = .equalSpacing
		toolbarMimicStack.spacing = 0.0
		toolbarMimicStack.layoutMargins = UIEdgeInsets(
			top: 0,
			left: isPortraitOrientation() ? 5 : 10,
			bottom: 0,
			right: isPortraitOrientation() ? 5 : 10
		)

		// Add stackview to toolbar
		navigationController!.view.addSubview(toolbarMimicStack)
		toolbarMimicStack.translatesAutoresizingMaskIntoConstraints = false

		// Anchor stackview to toolbar constraints
		toolbarMimicStack.leftAnchor.constraint(equalTo: navigationController!.toolbar.leftAnchor).isActive = true
		toolbarMimicStack.rightAnchor.constraint(equalTo: navigationController!.toolbar.rightAnchor).isActive = true
		toolbarMimicStack.topAnchor.constraint(equalTo: navigationController!.toolbar.topAnchor).isActive = true
		toolbarMimicStack.bottomAnchor.constraint(equalTo: navigationController!.toolbar.bottomAnchor).isActive = true

		// Add UIViews with intrinsicContentSizes of 44*44px into stackview
		toolbarMimicStack.addArrangedSubview(ViewController.makeRepresentativeMimicButton())
		toolbarMimicStack.addArrangedSubview(ViewController.makeRepresentativeMimicButton())
		toolbarMimicStack.addArrangedSubview(ViewController.makeRepresentativeMimicButton())
		toolbarMimicStack.addArrangedSubview(ViewController.makeRepresentativeMimicButton())
	}

    func makeDummyToolbarItem() -> UIBarButtonItem {
    	return UIBarButtonItem(title: "三", style: UIBarButtonItemStyle.plain, target: self, action: nil)
    }

	func isPortraitOrientation() -> Bool {
		return UIApplication.shared.statusBarOrientation.isPortrait
	}

	/** Make an IntrinsicallySizedView with blue background */
	func makeRepresentativeMimicButton() -> UIView {
	    let uiview: IntrinsicallySizedView = IntrinsicallySizedView()
	    uiview.backgroundColor = UIColor.blue
	    return uiview
	}

	/** UIView with 44*44px intrinsicContentSize */
	class IntrinsicallySizedView: UIView {
	    override open var intrinsicContentSize: CGSize {
	        return CGSize(width: 44, height: 44)
	    }
	}

    /** Method from ViewController class called upon change in orientation. */
    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)

        coordinator.animate(alongsideTransition:
        	{ context in
	            guard self.toolbarMimicStack != nil else { return print("toolbarMimicStack was nil!"); }
	            
	            var portraitIncoming: Bool = true
	            if(size.width > size.height){
	                print("Landscape incoming")
	                portraitIncoming = false
	            } else {
	                print("Portrait incoming")
	            }

	            self.toolbarMimicStack.layoutMargins = UIEdgeInsets(
	            	top: 0,
	            	left: portraitIncoming ? 5 : 10,
	            	bottom: 0,
	            	right: portraitIncoming ? 5 : 10
	        	)
	        },
	        completion: nil
        )
    }
}

extension ViewController: TNTutorialManagerDelegate {
    func tutorialWrapUp() {
        print("Tutorial wrapped up.")
        tm = nil
    }
    
    func tutorialMaxIndex() -> Int {
        return 1
    }
    
    func tutorialMasterView() -> UIView! {
        return navigationController!.view
    }
    
    func tutorialViews(toHighlight index: Int) -> [UIView]! {
        switch(index){
        case 0:
            return [toolbarMimicStack.arrangedSubviews[0], toolbarMimicStack.arrangedSubviews[1]]
        default:
            return nil
        }
    }

    func tutorialIdentifier() -> String! {
        return "mainVC"
    }
}
```

## Observed behaviour

The highlights will initially appear perfectly centred for the given rotation, but then will come out of line once rotated, because we update the stackview's margins in `viewWillTransition()`.

## Desired behaviour

The highlights should take the frame of the stackview's up-to-date positions upon each rotation. In fact, it would be very nice if they could even constantly track a target (like a fish that moves around the screen). I realise this is likely an advanced feature addition, however.

**Comment by Tawa** [comment]
Can you try to call `updateTutorial` after the frames change?

**Comment by shirakaba** [comment]
Hi Tawa, sorry for the late response – still very busy due to the Christmas season. Thanks for releasing an update for the latest pull request in the meantime. I see you've made a commit regarding an orientation change issue: https://github.com/Tawa/TNTutorialManager/commit/9e65afb2055e53a83601ca1c5a46dde667b35c15

Provided I get some time, I'll check whether that has any effect and also try calling `updateTutorial()` as you say. Can't say whether I'll manage it this weekend, however.

**Comment by shirakaba** [comment]
@Tawa Just beginning to have a look into this now. Looks like the issue is much more obvious on an iPad: forget about the highlight; even the shadow doesn't update to fit to the new screen size.

I tried calling `updateTutorial()` and it redrew exactly the same (incorrect) view. I wonder if it has `width` and `height` mixed up...

About to step through the code and see if I can learn something. Hopefully I get some time to take some action on it.

**Comment by shirakaba** [comment]
Okay, it appears that was just a distraction. Looks like your commit [9e65afb](https://github.com/Tawa/TNTutorialManager/commit/9e65afb2055e53a83601ca1c5a46dde667b35c15) fixes the shaded area upon iPad rotation, but doesn't have any impact on the issue in this discussion itself.

Sidenote: why did you reverse that commit? It's a good fix.

**Comment by shirakaba** [comment]
Okay, I've solved it. It looks like it was just a trivial race condition:

If, during your main app's `viewWillTransitionToSize()` block, you ask for a layout change using a `coordinator.animate()` closure, then of course those layout changes won't have been applied by the time that TNTutorialManager is prompted to re-invoke `setupLayout()`.

I'll close this issue, as my main focus (my perceived bug upon orientation change) has been resolved. Admittedly the opening post for this issue did also mention the possibility of updating the highlight position for a constantly-moving UIView, but that's clearly a feature request rather than a bug-fix, and as such would merit a separate issue being opened. In any case, I don't have such a use case, so I won't be opening such an issue for it in the foreseeable future.
